examples/schedulers/static_thread_pool.hpp

```diff
@@ -88,11 +89,275 @@ namespace example {
+            -> std::execution::env_of_t<Receiver> {
+            return std::execution::get_env(self.shared_state_.receiver_);
+          }
+        };
+
+      template <class SenderId, class ReceiverId, std::integral Shape, class Fun>
+        struct bulk_op_state {
+          using Sender = std::__t<SenderId>;
+          using Receiver = std::__t<ReceiverId>;
+
+          static constexpr bool may_throw =
+              !std::__v<std::execution::__value_types_of_t<
+                  Sender, std::execution::env_of_t<Receiver>,
+                  std::__mbind_front_q<bulk_non_throwing, Fun, Shape>,
+                  std::__q<std::__mand>>>;
+
+          using bulk_receiver = bulk_receiver<SenderId, ReceiverId, Shape, Fun, may_throw>;
+          using shared_state = bulk_shared_state<SenderId, ReceiverId, Shape, Fun, may_throw>;
+          using inner_op_state = std::execution::connect_result_t<Sender, bulk_receiver>;
+
+          shared_state shared_state_;
+
+          inner_op_state inner_op;
+
+          friend void tag_invoke(std::execution::start_t, bulk_op_state& op) noexcept {
+            std::execution::start(op.inner_op);
+          }
+
+          bulk_op_state(static_thread_pool &pool, Shape shape, Fun fn, Sender&& sender, Receiver receiver)
+            : shared_state_(pool, (Receiver&&)receiver, shape, fn)
+            , inner_op{std::execution::connect((Sender&&)sender, bulk_receiver{shared_state_})} {
+          }
+        };
+
+      template <class SenderId, std::integral Shape, class FunId>
+        struct bulk_sender {
+          using Sender = std::__t<SenderId>;
+          using Fun = std::__t<FunId>;
+
+          static_thread_pool& pool_;
+          Sender sndr_;
+          Shape shape_;
+          Fun fun_;
+
+          template <class Fun, class Sender, class Env>
+            using with_error_invoke_t =
+              std::__if_c<
+                std::__v<std::execution::__value_types_of_t<
+                  Sender,
+                  Env,
+                  std::__mbind_front_q<bulk_non_throwing, Fun, Shape>,
+                  std::__q<std::__mand>>>,
+                std::execution::completion_signatures<>,
+                std::execution::__with_exception_ptr>;
+
+          template <class Self, class Env>
+            using completion_signatures =
+              std::execution::__make_completion_signatures<
+                std::__member_t<Self, Sender>,
+                Env,
+                with_error_invoke_t<Fun, std::__member_t<Self, Sender>, Env>>;
+
+          template <class Self, class Receiver>
+            using bulk_op_state_t = 
+              bulk_op_state<
+                std::__x<std::__member_t<Self, Sender>>, 
+                std::__x<std::remove_cvref_t<Receiver>>, Shape, Fun>;
+
+          template <std::__decays_to<bulk_sender> Self, std::execution::receiver Receiver>
+            requires std::execution::sender_to<std::__member_t<Self, Sender>, Receiver>
```

Comment:
This constraint also doesn't factor in the fact that the `bulk_sender` decay-copies values. Maybe this instead (once you make the above change to `completion_signatures`):

```suggestion
            requires std::execution::receiver_of<Receiver, completion_signatures<Self, env_of_t<Receiver>>
```